DNN Platform/Library/Services/ImprovementsProgram/BeaconService.cs

```diff
@@ -0,0 +1,117 @@
+using System.Text;
+using System.Web;
+using DotNetNuke.Entities.Host;
+using DotNetNuke.Entities.Portals;
+using DotNetNuke.Entities.Tabs;
+using DotNetNuke.Entities.Users;
+using DotNetNuke.Framework;
+
+namespace DotNetNuke.Services.ImprovementsProgram
+{
+    public class BeaconService : ServiceLocator<IBeaconService, BeaconService>, IBeaconService
+    {
+        protected override Func<IBeaconService> GetFactory()
+        {
+            return () => new BeaconService();
+        }
+
+        private string _beaconEndpoint;
+        private readonly SHA256 _sha256 = SHA256.Create();
+
+        public string GetBeaconEndpoint()
+        {
+            if (string.IsNullOrEmpty(_beaconEndpoint))
+            {
+                var ep = ConfigurationManager.AppSettings["ImprovementProgram.Epdpoint"];
+#if DEBUG
+                _beaconEndpoint = string.IsNullOrEmpty(ep)
+                    ? "https://dev-bcn.dnnapi.com/beacon"
+                    : ep;
+#else
+                _beaconEndpoint = string.IsNullOrEmpty(ep)
+                    ? "https://bcn.dnnapi.com/beacon"
+                    : ep;
+#endif
+            }
+            return _beaconEndpoint;
+        }
+
+        public bool IsBeaconEnabledForControlBar(UserInfo user)
+        {
+            //check for Update Service Opt-in
+            //check if a host or admin
+            //check if currently on a host/admin page
+            var enabled = false;
+
+            if (Host.ParticipateInImprovementProg)
+            {
+                var roles = GetUserRolesBitValues(user);
+                var tabPath = TabController.CurrentPage.TabPath;
+                enabled = (roles & (RolesEnum.Host | RolesEnum.Admin)) != 0 && 
+                    (tabPath.StartsWith("//Admin") || tabPath.StartsWith("//Host"));
+            }
+
+            return enabled;
+        }
+
+        public string GetBeaconQuery(UserInfo user, string filePath = null)
+        {
+            var roles = 0;
+            if (user.UserID >= 0)
+            {
+                if (user.IsSuperUser) roles |= (int)RolesEnum.Host;
+                if (user.IsInRole("Administrators")) roles |= (int)RolesEnum.Admin;
+                if (user.IsInRole("Content Managers")) roles |= (int)RolesEnum.ContentManager;
+                if (user.IsInRole("Content Editors")) roles |= (int)RolesEnum.ContentEditor;
+                if (user.IsInRole("Community Manager")) roles |= (int)RolesEnum.CommunityManager;
+            }
+
+            // h: Host GUID - hashed
+            // p: Portal ID
+            // a: Portal Alias
+            // r: Role(s) - bitmask - see RolesEnum
+            // u: User ID - hashed
+            // f: filename/path (optional; unused here)
+
+            var portalSettings = PortalController.Instance.GetCurrentPortalSettings();
+            var qparams = new Dictionary<string, string>
+            {
+                // Remember to URL ENCODE values that can be ambigious
```

Comment:
why not just URL encode all values to make sure they're always safe (e.g. in the `Select` on line 95)?